Update of CRPS tutorial ready for review

DerynGriffiths opened the pull request:
This is not ready to merge. It has a cell with a question for Tennessee.

I have made substantial changes to the previous version as part of a review process, while keeping the same example.

Please provide comments.

nicholasloveday commented:
@DerynGriffiths, nice start.
There is also CRPS for ensembles https://github.com/nci/scores/blob/develop/src/scores/probability/crps_impl.py#L749

It would be good to demonstrate how the threshold weighted version works too

DerynGriffiths commented:
I've made changes following review by @nicholasloveday.

tennlee commented:
This is obviously better than what was there before, but by virtue of being better and more detail, also prompted more feedback and thought on my part... 

Should we actually label/number the figures? I'm not really sure.

Reflecting, if this was actually 1 minus the CDF and the "chance or reaching" it's a more intuitive explanation to me. Presenting both charts might help in aiding understanding.

With the observation section, technically there is error in the observation as well, to the calibration and accuracy of the sensor. I don't know whether it's significant, but also definitively saying there is no chance of observing less than the observed value might not be the whole story.

The chart of the CRPS components is looking great.

The example of when to use threshold_weight and how deserves a little more exposition. Why is it reasonable to weight both fcst and obs below 4 in that instance? It's not necessarily obvious that 4 degrees has anything to do with a severe frost - wouldn't like minus 5 be more severe? I realise the example is based on setting aside obviously warmer temperatures, but even so some kind of definition of a severe frost might be helpful here. 

I think the CRPS for ensemble needs a bit more introduction. I would consider separating it into a separate notebook, or expanding on how what ensemble information even is. I realise it's said - equally likely forecasts. But it may not be obvious where the CDF from the previous example came from, if not an ensemble, and why the equally likely forecasts don't have CDFs of their own. Maybe saying equally likely deterministic forecasts would be sufficient. Maybe plot the ensemble information to show how it relates to the bell curve of the initial probablistic forecast used earlier. 

Let's talk later about how to distil this into some more actionable feedback, but this is all my initial thoughts and perhaps the detail is helpful.

nicholasloveday commented:
A few thoughts:
- Ln [9]. Add a sentence to explain what the `threshold_dim` arg does
- In the `Things to try next` section, convert the list to use dot points
- You could also suggest in the `Things to try next` section, that the user plays around with the weights so that they are not just binary. I'm fairly happy with your section on using 4C as the frost threshold, but perhaps if there are some other thresholds (less than 4C) that are even more important, then they could be weighted more heavily. I think we do need an explanation of how threshold weighting works (you could even do a shaded area for twCRPS plot) though . 
- For the `crps_for_ensemble` you can mention that there are two methods (empirical and Fair). You could consider giving a short paragraph, perhaps with the equation for each, and briefly explain the differences between the two methods. If you don't, then you should at least add the Ferro paper to your final dotpoint in the things to try next https://rmets.onlinelibrary.wiley.com/doi/10.1002/qj.2270

> Reflecting, if the figure showing "chance of not exceeding" was actually 1 minus the CDF and the "chance or reaching" it's a more intuitive explanation to me. Presenting both charts might help in aiding understanding.

I'm comfortable with just showing "Chance of non-exceedance". I think it should be "exceedance" not "exceedence" in the plot title

nicholasloveday commented:
> Maybe plot the ensemble information to show how it relates to the bell curve of the initial probabilistic forecast used earlier.

I think this could be a good idea

DerynGriffiths commented:
I have considered all comments and acted on many. In particular I have put the crps_for_ensemble in a separate tutorial and shown an empircial cdf for the example. 
Please review again.

tennlee commented:
In the notebook I note you say "The larger the ensemble the smoother and more sensible this will be". I have no idea what modes of variability exist in ensembles and whether you end up with severely bimodal distributions or significant discontinuities in practise. If you don't see such discontinuities, what is the value of the deterministic ensemble as opposed to just carrying a variance around?

I really like that bit of the notebook generally, I just don't know if that particular comment is always the case or not.

It would be nice to see the area-representation of the CRPS again if that's reasonable to suggest. It's implied, but it might be nice to see the plot anyway.

Looks great overall, very happy, thank you

Please add the new notebooks to the "Explanation" notebook so they appear in the documentation

DerynGriffiths commented:
I've added the area-representation to the second CRPS notebook.
I've added a link from the "Explanation" notebook. Please check the labelling for me.
I think this could be merged.
